src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -112496,5 +112489,186 @@ void M()
+        _ = i4;
+
+        IEquatable<(B?, B?)> i5 = new B();
+        IEquatable<(B, B)> i6 = i5;
+        IEquatable<(B, B)> i7 = new B();
+        IEquatable<(B?, B?)> i8 = i7; // warn
+        _ = i8;
+    }
+}",
+                new[] { spanRef },
+                options: WithNonNullTypesTrue().WithSpecificDiagnosticOptions("CS0436", ReportDiagnostic.Suppress));
+            comp.VerifyDiagnostics(
+                // (34,29): warning CS8619: Nullability of reference types in value of type 'IEquatable<A>' doesn't match target type 'IEquatable<A?>'.
+                //         IEquatable<A?> i4 = i3; // warn
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "i3").WithArguments("System.IEquatable<A>", "System.IEquatable<A?>").WithLocation(41, 29),
+                // (40,35): warning CS8619: Nullability of reference types in value of type 'IEquatable<(B, B)>' doesn't match target type 'IEquatable<(B?, B?)>'.
+                //         IEquatable<(B?, B?)> i8 = i7; // warn
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "i7").WithArguments("System.IEquatable<(B, B)>", "System.IEquatable<(B?, B?)>").WithLocation(47, 35)
+                );
+
+            // Test with non-wellknown type
+            var defComp = CreateCompilation(def,
+                references: new[] { spanRef },
+                options: WithNonNullTypesTrue().WithSpecificDiagnosticOptions("CS0436", ReportDiagnostic.Suppress));
+            defComp.VerifyDiagnostics();
+            var useComp = CreateCompilation(@"
+using System;
+public interface IEquatable<T>
+{
+    bool Equals(T other);
+}
+class C
+{
+    static void Main()
+    {
+        var x = new Span<A?>();
+        var y = new Span<A>();
+
+        A.M(x);
+        A.M(y);
+    }
+}",
+                references: new[] { spanRef, defComp.ToMetadataReference() },
+                options: WithNonNullTypesTrue());
+            useComp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void IEquatableNotContravariantExceptNullability()
+        {
+            var src = @"
+using System;
+
+class A : IEquatable<A?>
+{
+    public bool Equals(A? a) => false;
+}
+
+class B : A {}
+
+class C
+{
+    static void Main()
+    {
+        var x = new Span<B?>();
+        var y = new Span<B>();
+
+        M(x);
+        M(y);
+    }
+    static void M<T>(Span<T> s) where T : IEquatable<T>?  { }
+}";
+            var comp = CreateCompilationWithSpan(src + @"
+namespace System
+{
+    public interface IEquatable<T>
+    {
+        bool Equals(T other);
+    }
+}
+", options: WithNonNullTypesTrue().WithSpecificDiagnosticOptions("CS0436", ReportDiagnostic.Suppress));
+            comp.VerifyDiagnostics(
+                // (25,9): error CS0311: The type 'B' cannot be used as type parameter 'T' in the generic type or method 'C.M<T>(Span<T>)'. There is no implicit reference conversion from 'B' to 'System.IEquatable<B>'.
+                //         M(x);
```

Comment:
it would help readability of the test if you added `// 1`, `// 2` etc comments on the lines which are expected to produce diagnostics.

Reply:
In this case there are only two method calls to M in the program and both of them have errors, but I'll try to do that more in the future.